File: exercise/C#/day24/tests/Delivery.Tests/Doubles/InMemoryToyRepository.cs
using Delivery.Domain;
using Delivery.Domain.Core;
using LanguageExt;
using static LanguageExt.Option<Delivery.Domain.Toy>;

namespace Delivery.Tests.Doubles
{
    public class InMemoryToyRepository : IToyRepository
    {
        private Map<Guid, Toy> _toys;
        private Seq<IEvent> _raisedEvents;

        public Option<Toy> FindByName(string toyName)
            => _toys.Filter(toy => toy.Name == toyName)
                .Values
                .ToOption();

        public Option<Toy> FindById(Guid id)
            => _toys.ContainsKey(id)
                ? _toys[id]
                : None;

        public void Save(Toy toy)
        {
            _raisedEvents = [];
            _toys = _toys.AddOrUpdate(toy.Id, toy);

            ((IAggregate) toy).GetUncommittedEvents()
                .ToList()
                .ForEach(@event => { _raisedEvents = _raisedEvents.Add(@event); });

            ((IAggregate) toy).ClearUncommittedEvents();
        }

        public Seq<IEvent> RaisedEvents() => _raisedEvents;
    }
}

File: exercise/ts/day24/tests/doubles/InMemoryToyRepository.ts
import {none, Option, some} from 'fp-ts/Option';
import {IToyRepository} from "../../src/domain/IToyRepository";
import {Toy} from "../../src/domain/Toy";

import {IEvent} from "../../src/domain/core/IEvent";

export class InMemoryToyRepository implements IToyRepository {
    private _toys: Map<string, Toy> = new Map<string, Toy>();
    private _raisedEvents: IEvent[] = [];

    findByName(name: string): Option<Toy> {
        for (let toy of this._toys.values()) {
            if (toy.name === name) {
                return some(toy);
            }
        }
        return none;
    }

    findById(id: string): Option<Toy> {
        const toy = this._toys.get(id);
        return toy ? some(toy) : none;
    }

    save(toy: Toy): void {
        this._raisedEvents = [];
        this._toys = this._toys.set(toy.getId(), toy);

        toy.getUncommittedEvents().forEach((event) => {
            this._raisedEvents.push(event);
        });

        toy.clearUncommittedEvents();
    }

    raisedEvents(): IEvent[] {
        return this._raisedEvents;
    }
}